src/include/duckdb/main/secret/duck_secret_manager.hpp

```diff
@@ -0,0 +1,106 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/main/secret/duck_secret_manager.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/catalog/default/default_generator.hpp"
+#include "duckdb/common/common.hpp"
+#include "duckdb/main/secret/secret.hpp"
+#include "duckdb/main/secret/secret_manager.hpp"
+
+namespace duckdb {
+class SecretManager;
+struct DBConfig;
+class SchemaCatalogEntry;
+
+//! The Main secret manager implementation for DuckDB. The secret manager can handle both temporary and permanent
+//! secrets
+class DuckSecretManager : public SecretManager {
+	friend struct SecretEntry;
+
+public:
+	explicit DuckSecretManager() = default;
+	virtual ~DuckSecretManager() override = default;
+
+	//! SecretManager API
+	DUCKDB_API void Initialize(DatabaseInstance &db) override;
+	DUCKDB_API unique_ptr<BaseSecret> DeserializeSecret(CatalogTransaction transaction,
+	                                                    Deserializer &deserializer) override;
+	DUCKDB_API void RegisterSecretType(CatalogTransaction transaction, SecretType &type) override;
+	DUCKDB_API SecretType LookupType(CatalogTransaction transaction, const string &type) override;
+	DUCKDB_API void RegisterSecretFunction(CatalogTransaction transaction, CreateSecretFunction function,
+	                                       OnCreateConflict on_conflict) override;
+	DUCKDB_API optional_ptr<SecretEntry> RegisterSecret(CatalogTransaction transaction,
+	                                                    unique_ptr<const BaseSecret> secret,
+	                                                    OnCreateConflict on_conflict,
+	                                                    SecretPersistMode persist_mode) override;
+	DUCKDB_API optional_ptr<SecretEntry> CreateSecret(ClientContext &context, const CreateSecretInfo &info) override;
+	DUCKDB_API BoundStatement BindCreateSecret(CatalogTransaction transaction, CreateSecretStatement &stmt) override;
+	DUCKDB_API optional_ptr<SecretEntry> GetSecretByPath(CatalogTransaction transaction, const string &path,
+	                                                     const string &type) override;
+	DUCKDB_API optional_ptr<SecretEntry> GetSecretByName(CatalogTransaction transaction, const string &name) override;
+	DUCKDB_API void DropSecretByName(CatalogTransaction transaction, const string &name, bool missing_ok) override;
+	DUCKDB_API vector<SecretEntry *> AllSecrets(CatalogTransaction transaction) override;
+	DUCKDB_API bool AllowConfigChanges() override;
+
+	//! Return secret directory
+	string GetSecretDirectory(DBConfig &config);
```

Comment:
note: this doesn't need to be public we can pass this on creation to the default generator